### ISSUE TITLE
CUDA out of memory while finetuning GPT

### ISSUE THREAD

**Issue by StrangeTcy** [microsoft/DeepSpeed#1276]
On Ubuntu 18.04, using deepspeed==0.4.4 and the following script:

```
deepspeed --include localhost:0,1,2,3   transformers/examples/pytorch/language-modeling/run_clm.py \
       --model_name_or_path EleutherAI/gpt-neo-2.7B \
      --train_file  /home/ubuntu/part0038.txt\
       --validation_file /home/ubuntu/part0037.txt \
       --do_train \
       --do_eval \
       --per_device_train_batch_size 2 \
       --logging_dir our_logs_gptneo \
       --deepspeed our_gpt3_large_2048.json \
       --seed $SEED \
       --fp16 \
       --fp16_backend auto \
       --run_name ru_gptneo \
       --report_to all \
       --save_step 1000 \
       --save_total_limit 5 \
       --output_dir ru_gptneo/model;
```

I run into the `CUDA out of memory` error, stating that 100 MB failed to get allocated. This was tried on different machines, including those with 16 and 32 GB of GPU memory, so I  really doubt that it's a lack of memory issue.
The gptneo model is 10.7 GB, which should fit into a single GPU memory.

**Comment by tjruwase** [comment]
@StrangeTcy, thanks for reporting this issue. Can you please share your deepspeed config json?

**Comment by StrangeTcy** [comment]
```
{
  "train_micro_batch_size_per_gpu": 2,
  "gradient_accumulation_steps":1,
  "steps_per_print":100,
  "gradient_clipping":1,
  "fp16": {
    "enabled": true,
    "loss_scale": 0,
    "loss_scale_window": 2000,
    "hysteresis":2,
    "min_loss_scale": 0.0
  },
  "zero_optimization": {
    "stage": 0,
    "reduce_bucket_size": 50000000
  }
}

```

**Comment by tjruwase** [comment]
Thanks for the sharing the deepspeed config. I notice that you re not using the ZeRO optimizer, which are enabled by stages 1, 2, & 3.

> The gptneo model is 10.7 GB, which should fit into a single GPU memory.
Can you clarify this a bit. Do you mean the checkpoint file on disk is 10.7GB? Does that include optimizer state? It would be better if you could provide the number of parameters in the model. That will help to better estimate the overall memory consumption of the model and optimizer states.

**Comment by StrangeTcy** [comment]
> Thanks for the sharing the deepspeed config. I notice that you re not using the ZeRO optimizer, which are enabled by stages 1, 2, & 3.
Thanks, I probably should try to.

> 
> > The gptneo model is 10.7 GB, which should fit into a single GPU memory.
> > Can you clarify this a bit. Do you mean the checkpoint file on disk is 10.7GB? Does that include optimizer state? It would be better if you could provide the number of parameters in the model. That will help to better estimate the overall memory consumption of the model and optimizer states.

What I mean is this: the `pytorch_model.bin` file has a 10.7 GB size, at least on the huggingface site it does (https://huggingface.co/EleutherAI/gpt-neo-2.7B/tree/main). 
As for the number of parameters, I don't really remember it and it's not readily available, but my guess is that `gpt-neo-2.7B` implies 2.7 billion parameters or thereabouts.

**Comment by StrangeTcy** [comment]
Okay, adding the ZeRO optimisation key as per [the tutorial](https://www.deepspeed.ai/tutorials/zero/) allowed it to start training. Thanks for the pointers.
